### PR TITLE
Fix #3335 - filteredTableData now syncs with tableData

### LINKED ISSUE #3335: [Bug] Table selectedRow breaks when a new record is added to the table data
## Description

The standup app selected row broke once a new record was added in the middle of the table.

### PR DESCRIPTION
**Pull request by riodeuno**
## Description
Make `filteredTableData` and computed `columns` as derived properties

Fixes #3335

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

All existing tests must pass

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/widgets/TableWidget/derived.js [hunk 1]
```diff
@@ -92,4 +92,299 @@ export default {
     return [];
   },
   //
+  getTableColumns: (props) => {
```
Comment (hetunandu) [inline app/client/src/widgets/TableWidget/derived.js:95]: Can we add some jest test for this?

Comment (hetunandu) [inline app/client/src/widgets/TableWidget/derived.js:95]: Wondering why did we not make this file a typescript file

File: app/client/src/widgets/TableWidget/index.tsx [hunk 2]
```diff
@@ -826,10 +638,14 @@ class TableWidget extends BaseWidget<TableWidgetProps, WidgetState> {
               : this.props.selectedRowIndex
           }
           multiRowSelection={this.props.multiRowSelection}
-          selectedRowIndices={computedSelectedRowIndices}
-          serverSidePaginationEnabled={serverSidePaginationEnabled}
+          selectedRowIndices={
+            Array.isArray(this.props.selectedRowIndices)
+              ? this.props.selectedRowIndices
+              : []
```
Comment (hetunandu) [inline app/client/src/widgets/TableWidget/index.tsx:644]: Should be handled by the widget validations

File: app/client/src/widgets/TableWidget/index.tsx [hunk 3]
```diff
@@ -826,10 +638,14 @@ class TableWidget extends BaseWidget<TableWidgetProps, WidgetState> {
...
+          }
+          serverSidePaginationEnabled={!!this.props.serverSidePaginationEnabled}
           onRowClick={this.handleRowClick}
-          pageNo={pageNo}
+          pageNo={this.props.pageNo || 1}
```
Comment (hetunandu) [inline app/client/src/widgets/TableWidget/index.tsx:648]: Should be handled by the widget validations

**Review by hetunandu — CHANGES_REQUESTED** [review]
Really happy to see become a derived property. Table is less complex today thanks to this change 😁

File: app/client/src/widgets/TableWidget/derived.js (hunk 1 above)
Comment (riodeuno) [reply app/client/src/widgets/TableWidget/derived.js:95]: This is technically loading this file as a raw string and then parsing the different functions in the `parsedDerivedProperties.ts`. These functions are then passed as is in the derivedPropertiesMap.

Comment (riodeuno) [reply app/client/src/widgets/TableWidget/derived.js:95]: Yes, these are designed this way - pure functions using only props as an argument for this very reason. Complicated derived properties can now be tested avoiding issues we faced in the TextWidget.

Comment (riodeuno) [reply app/client/src/widgets/TableWidget/derived.js:95]: We should also make the `parsedDerivedProperties.ts` a generic utility function to be used in other widgets. This is something we're getting done in the list widget project.

**Review by hetunandu — DISMISSED** [review]

**Review by areyabhishek — APPROVED** [review]

**Pull request merged**